src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs

```diff
@@ -223,88 +221,56 @@ internal void ForceCancelRequest()
-                    requestQueueHandle: Server.RequestQueue.Handle,
-                    requestId,
-                    propertyId: (HTTP_REQUEST_PROPERTY)11 /* HTTP_REQUEST_PROPERTY.HttpRequestPropertyTlsClientHello  */,
-                    qualifier: null,
-                    qualifierSize: 0,
-                    output: pBuffer,
-                    outputSize: (uint)buffer.Length,
-                    bytesReturned: (IntPtr)bytesReturned,
-                    overlapped: IntPtr.Zero);
-
-                if (statusCode is ErrorCodes.ERROR_SUCCESS)
-                {
-                    tlsClientHelloBytesCallback(features, buffer.AsSpan(0, (int)bytesReturnedValue));
-                    return true;
-                }
-            }
-        }
-        finally
-        {
-            ArrayPool<byte>.Shared.Return(buffer, clearArray: true);
-        }
+        uint bytesReturnedValue = 0;
+        uint* bytesReturnedPointer = &bytesReturnedValue;
 
-        // if buffer supplied is too small, `bytesReturned` will have proper size
-        // so retry should succeed with the properly allocated buffer
-        if (statusCode is ErrorCodes.ERROR_MORE_DATA or ErrorCodes.ERROR_INSUFFICIENT_BUFFER)
+        fixed (byte* pBuffer = destination)
         {
-            try
-            {
-                var correctSize = (int)bytesReturnedValue;
-                buffer = ArrayPool<byte>.Shared.Rent(correctSize);
+            statusCode = HttpApi.HttpGetRequestProperty(
+                requestQueueHandle: Server.RequestQueue.Handle,
+                requestId,
+                propertyId: (HTTP_REQUEST_PROPERTY)11 /* HTTP_REQUEST_PROPERTY.HttpRequestPropertyTlsClientHello  */,
+                qualifier: null,
+                qualifierSize: 0,
+                output: pBuffer,
+                outputSize: (uint)destination.Length,
+                bytesReturned: (IntPtr)bytesReturnedPointer,
+                overlapped: IntPtr.Zero);
 
-                fixed (byte* pBuffer = buffer)
-                {
-                    statusCode = HttpApi.HttpGetRequestProperty(
-                        requestQueueHandle: Server.RequestQueue.Handle,
-                        requestId,
-                        propertyId: (HTTP_REQUEST_PROPERTY)11 /* HTTP_REQUEST_PROPERTY.HttpRequestPropertyTlsClientHello  */,
-                        qualifier: null,
-                        qualifierSize: 0,
-                        output: pBuffer,
-                        outputSize: (uint)buffer.Length,
-                        bytesReturned: (IntPtr)bytesReturned,
-                        overlapped: IntPtr.Zero);
-
-                    if (statusCode is ErrorCodes.ERROR_SUCCESS)
-                    {
-                        tlsClientHelloBytesCallback(features, buffer.AsSpan(0, correctSize));
-                        return true;
-                    }
-                }
+            bytesReturned = (int)bytesReturnedValue;
```

Comment:
Are we sure this can't overflow?

Reply:
I think we are fine here, because TLS client hello has to be 2048 MB to have overflow here. If there is a TLS client hello of 2 GB then there is definitely something more wrong here?